endpoints/openrtb2/video_auction.go

```diff
@@ -0,0 +1,523 @@
+		handleError(labels, w, errL, ao)
+		return
+	}
+	bidReq.Imp = imps
+	bidReq.ID = "bid_id" //TODO: look at prebid.js
+
+	// Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
+	deps.setFieldsImplicitly(r, bidReq) // move after merge
+
+	errL = deps.validateRequest(bidReq)
+	if len(errL) > 0 {
+		handleError(labels, w, errL, ao)
+		return
+	}
+
+	ctx := context.Background()
+	cancel := func() {}
+	timeout := deps.cfg.AuctionTimeouts.LimitAuctionTimeout(time.Duration(bidReq.TMax) * time.Millisecond)
+	if timeout > 0 {
+		ctx, cancel = context.WithDeadline(ctx, start.Add(timeout))
+	}
+	defer cancel()
+
+	usersyncs := usersync.ParsePBSCookieFromRequest(r, &(deps.cfg.HostCookie))
+	if bidReq.App != nil {
+		labels.Source = pbsmetrics.DemandApp
+	} else {
+		labels.Source = pbsmetrics.DemandWeb
+		if usersyncs.LiveSyncCount() == 0 {
+			labels.CookieFlag = pbsmetrics.CookieFlagNo
+		} else {
+			labels.CookieFlag = pbsmetrics.CookieFlagYes
+		}
+	}
+
+	numImps = len(bidReq.Imp)
+
+	//execute auction logic
+	response, err := deps.ex.HoldAuction(ctx, bidReq, usersyncs, labels, &deps.categories)
+	ao.Request = bidReq
+	ao.Response = response
+	if err != nil {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+
+	//build simplified response
+	bidResp, err := buildVideoResponse(response)
+	if err != nil {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+	if bidReq.Test == 1 {
+		bidResp.Ext = response.Ext
+	}
+
+	resp, err := json.Marshal(bidResp)
+	//resp, err := json.Marshal(response)
+	if err != nil {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(resp)
+
+}
+
+func handleError(labels pbsmetrics.Labels, w http.ResponseWriter, errL []error, ao analytics.AuctionObject) {
+	labels.RequestStatus = pbsmetrics.RequestStatusErr
+	w.WriteHeader(http.StatusInternalServerError)
+	var errors string
+	for _, er := range errL {
+		errors = fmt.Sprintf("%s %s", errors, er.Error())
+	}
+	fmt.Fprintf(w, "Critical error while running the video endpoint: %v", errors)
+	glog.Errorf("/openrtb2/video Critical error: %v", errors)
+	ao.Status = http.StatusInternalServerError
+	ao.Errors = append(ao.Errors, errL...)
+}
+
+func (deps *endpointDeps) createImpressions(videoReq *openrtb_ext.BidRequestVideo) (imps []openrtb.Imp, err []error) {
+	videoDur := videoReq.PodConfig.DurationRangeSec
+	minDuration, maxDuration := minMax(videoDur)
+	reqExactDur := videoReq.PodConfig.RequireExactDuration
+	videoData := videoReq.Video
+
+	finalImpsArray := make([]openrtb.Imp, 0)
+	for _, pod := range videoReq.PodConfig.Pods {
+
+		//load stored impression
+		storedImpressionId := string(pod.ConfigId)
+		storedImp, errs := deps.loadStoredImp(storedImpressionId)
+		if errs != nil {
+			return nil, errs
+		}
+
+		numImps := pod.AdPodDurationSec / minDuration
+
+		impDivNumber := numImps / len(videoDur)
```

Comment:
If number of impressions (numImps) is less than the size of videoDur, we get impDivNumber == 0.  That is a possibility, I think.  When we divide by impDivNumber on line 263, something bad probably happens.  If so, we probably need to recognize a zero value here and correct it to one.

I suppose it might not be a bad idea to also check that minDuration is non-zero since we divide by that on line 253.

Reply:
`If number of impressions (numImps) is less than the size of videoDur` - I think this is exceptional and edge case. We need to discuss this. Which ad duration to use in this case.

Reply:
Thanks @josephveach . Good catch. For now, let's throw an error here if such a case arises.
As an incremental approach we need to add better validation on the incoming request. Also while doing so i will sync it with what we have on the client side.

@Kalypsonika Veronika will go ahead and add the error so that we don't get into a bad state.

Reply:
I'm not sure this should necessarily be an error.  Example:  AdPodDurationSec = 60, videoDur = (60, 40, 30, 20).  In this case, minDuration will be 20 and numImps will be 3.  Since len(videoDur) is 4, impDivNumber will be zero.  Why not simply correct impDivNumber to 1 here?

Reply:
Or should we increase the number of imps to ensure that `impDivNumbere` is at least one? This will give us an imp for each duration.

Reply:
Done